src/platform/PICO/uart/uart_pio.c

```diff
@@ -0,0 +1,374 @@
+    PIO_INTR_SM0_RXNEMPTY_BITS,
+    PIO_INTR_SM1_RXNEMPTY_BITS,
+    PIO_INTR_SM2_RXNEMPTY_BITS,
+    PIO_INTR_SM3_RXNEMPTY_BITS,
+};
+
+static const uint32_t txnfullbit[4] = {
+    PIO_INTR_SM0_TXNFULL_BITS,
+    PIO_INTR_SM1_TXNFULL_BITS,
+    PIO_INTR_SM2_TXNFULL_BITS,
+    PIO_INTR_SM3_TXNFULL_BITS,
+};
+
+// PIO-based UARTs. For now, hardwired to PIOUARTs 0,1 on PIO number UART_PIO_INDEX.
+const pioUartHardware_t pioUartHardware[SERIAL_PIOUART_COUNT] = {
+#ifdef USE_PIOUART0
+    {
+        .identifier = SERIAL_PORT_PIOUART0,
+        .irqn = PIO_IRQ_NUM(uartPio, 0),
+        .txBuffer = uartPio0TxBuffer,
+        .rxBuffer = uartPio0RxBuffer,
+        .txBufferSize = sizeof(uartPio0TxBuffer),
+        .rxBufferSize = sizeof(uartPio0RxBuffer),
+    },
+#endif
+
+#ifdef USE_PIOUART1
+    {
+        .identifier = SERIAL_PORT_PIOUART1,
+        .irqn = PIO_IRQ_NUM(uartPio, 1),
+        .txBuffer = uartPio1TxBuffer,
+        .rxBuffer = uartPio1RxBuffer,
+        .txBufferSize = sizeof(uartPio1TxBuffer),
+        .rxBufferSize = sizeof(uartPio1RxBuffer),
+    },
+#endif
+};
+
+static uartPinDef_t makePinDef(ioTag_t tag)
+{
+    uartPinDef_t ret = { .pin = tag };
+    return ret;
+}
+
+void uartPinConfigure_pio(const serialPinConfig_t *pSerialPinConfig)
+{
+    // software UART by PIO
+    int pinIndexMin = 48;
+    int pinIndexMax = -1;
+    uartPioBase = 0;
+    for (const pioUartHardware_t* hardware = pioUartHardware; hardware < ARRAYEND(pioUartHardware); hardware++) {
+        const serialPortIdentifier_e identifier = hardware->identifier;
+        uartDevice_t* uartdev = uartDeviceFromIdentifier(identifier);
+        const int resourceIndex = serialResourceIndex(identifier);
+        const ioTag_t cfgRx = pSerialPinConfig->ioTagRx[resourceIndex];
+        const ioTag_t cfgTx = pSerialPinConfig->ioTagTx[resourceIndex];
+        bprintf("pico uartPinConfigure pio at %p dev = %p,  tags rx 0x%x, tx 0x%x", hardware, uartdev, cfgRx, cfgTx);
+        if (!cfgRx && !cfgTx) {
+            continue;
+        }
+
+        // On a single PIO block, we are restricted either to pins 0-31 or pins 16-47.
+        pinIndexMin = cfgRx && (DEFIO_TAG_PIN(cfgRx) < pinIndexMin) ? DEFIO_TAG_PIN(cfgRx) : pinIndexMin;
+        pinIndexMax = cfgRx && (DEFIO_TAG_PIN(cfgRx) > pinIndexMax) ? DEFIO_TAG_PIN(cfgRx) : pinIndexMax;
+        pinIndexMin = cfgTx && (DEFIO_TAG_PIN(cfgTx) < pinIndexMin) ? DEFIO_TAG_PIN(cfgTx) : pinIndexMin;
+        pinIndexMax = cfgTx && (DEFIO_TAG_PIN(cfgTx) > pinIndexMax) ? DEFIO_TAG_PIN(cfgTx) : pinIndexMax;
+        if (pinIndexMax >= 32) {
+            if (pinIndexMin < 16) {
+                bprintf("* Not configuring PIOUART with identifier %d (PIO can't span pins min %d max %d)",
+                        identifier, pinIndexMin, pinIndexMax);
+                continue;
+            } else {
+                uartPioBase = 16;
+            }
+        }
+
+        if (cfgRx) {
+            uartdev->rx = makePinDef(cfgRx);
```

Comment:
Maybe (it may be easier to understand):
```suggestion
            uartdev->rx = (uartPinDef_t){ .pin =cfgRx};
```